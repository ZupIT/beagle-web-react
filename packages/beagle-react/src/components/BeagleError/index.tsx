/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { FC, useState } from 'react'
import { BeagleErrorInterface } from 'common/models'
import BeagleText from '../BeagleText'
import withTheme from '../utils/withTheme'
import {
  ErrorsDetailedContainer,
  StyledRetryButton,
  StyledRetryContainer,
  StyledShowMoreButton,
} from './styled'

const BeagleError: FC<BeagleErrorInterface> = ({
  retry,
  errors,
  className,
}) => {
  const [showingDetails, setShowingDetails] = useState<boolean>(false)

  const handleShowMore = () => {
    setShowingDetails(!showingDetails)
  }

  return (
    <StyledRetryContainer className={className}>
      <BeagleText text="Sorry!" textColor="red" />
      <BeagleText text="An unexpected error happened while loading your page." />
      <StyledRetryButton
        data-testid="retry"
        onPress={retry}
        text="Retry"
      ></StyledRetryButton>
      {errors.length > 0 ? (
        <>
          <StyledShowMoreButton
            data-testid="show-more"
            onPress={handleShowMore}
            text={`${showingDetails ? 'Hide' : 'Show'} details ${
              showingDetails ? '▲' : '▼'
            }`}
          ></StyledShowMoreButton>
          <ErrorsDetailedContainer
            data-testid="details"
            className={showingDetails ? 'show' : ''}
          >
            {errors.map((error, index) => (
              <p key={`key_${index}`} className="error-message">
                <span className="error-indicator">►</span> ERROR:{' '}
                {error.message}
              </p>
            ))}
          </ErrorsDetailedContainer>
        </>
      ) : null}
    </StyledRetryContainer>
  )
}

export default withTheme(BeagleError)
