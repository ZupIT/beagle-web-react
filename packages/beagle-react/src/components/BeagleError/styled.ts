/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import styled from 'styled-components'
import BeagleButton from '../BeagleButton'

export const StyledRetryContainer = styled.section`
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  margin: 2em;
  font-size: 23px;
`

export const StyledRetryButton = styled(BeagleButton)`
  max-width: 200px;
  margin: 8px 4px 4px 4px;
`

export const StyledShowMoreButton = styled(BeagleButton)`
  max-width: 200px;
  margin: 4px;
  border: none;
`

export const ErrorsDetailedContainer = styled.section`
  width: 100%;
  max-width: 800px;
  max-height: 0;
  opacity: 0;
  overflow: auto;
  position: relative;
  margin: 4px;
  border-radius: 8px;
  border: 1px solid #FFD7D7;
  background-color: #FFF0F0;
  padding: 8px;
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: flex-start;
  gap: 8px;
  transition: all 0.2s ease;

  &.show {
    max-height: 1024px;
    opacity: 1;
  }

  .error-message {
    font-family: Verdana, Geneva, Tahoma, sans-serif;
    position: relative;
    color: #FF0000;
    font-size: 14px;
    padding: 0;
    margin: 0;

    .error-indicator {
        color: #414a4c;
        margin: 0 4px 0 0;
        font-size: 12px;
    }
  }
`
