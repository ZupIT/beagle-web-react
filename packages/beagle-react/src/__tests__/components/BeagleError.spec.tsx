/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

// Link.react.test.js
import React from 'react'
import Adapter from 'enzyme-adapter-react-16'
import { configure, shallow } from 'enzyme'
import BeagleError from '../../components/BeagleError'

let wrapper: any


configure({ adapter: new Adapter() })
beforeAll(() => {
  wrapper = shallow(<BeagleError className="Test Class"/>)
})

test('Beagle snapshot Error', () => {
  expect(wrapper).toMatchSnapshot()
})

