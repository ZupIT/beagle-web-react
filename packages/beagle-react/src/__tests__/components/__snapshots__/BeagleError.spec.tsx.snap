// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Beagle Error - render correctly with details Should match the snapshot 1`] = `
<ComponentWithTheme
  className="Test Class"
  errors={
    Array [
      Object {
        "message": "First Message",
      },
      Object {
        "message": "Second Message",
      },
    ]
  }
  retry={[MockFunction]}
>
  <BeagleError
    className="Test Class"
    errors={
      Array [
        Object {
          "message": "First Message",
        },
        Object {
          "message": "Second Message",
        },
      ]
    }
    retry={[MockFunction]}
  >
    <styled.section
      className="Test Class"
    >
      <section
        className="sc-dlfnbm DatSL Test Class"
      >
        <ComponentWithTheme
          text="Sorry!"
          textColor="red"
        >
          <BeagleText
            className=""
            text="Sorry!"
            textColor="red"
          >
            <styled.p
              className=""
              textColor="red"
            >
              <p
                className="sc-bdfBwQ eliSds"
              >
                Sorry!
              </p>
            </styled.p>
          </BeagleText>
        </ComponentWithTheme>
        <ComponentWithTheme
          text="An unexpected error happened while loading your page."
        >
          <BeagleText
            className=""
            text="An unexpected error happened while loading your page."
          >
            <styled.p
              className=""
            >
              <p
                className="sc-bdfBwQ fTPwPK"
              >
                An unexpected error happened while loading your page.
              </p>
            </styled.p>
          </BeagleText>
        </ComponentWithTheme>
        <Styled(ComponentWithTheme)
          data-testid="retry"
          onPress={[MockFunction]}
          text="Retry"
        >
          <ComponentWithTheme
            className="sc-hKgILt lhgqZA"
            data-testid="retry"
            onPress={[MockFunction]}
            text="Retry"
          >
            <BeagleButton
              className="sc-hKgILt lhgqZA"
              data-testid="retry"
              onPress={[MockFunction]}
              text="Retry"
            >
              <styled.button
                aria-disabled={false}
                className="sc-hKgILt lhgqZA"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                <button
                  aria-disabled={false}
                  className="sc-gsTCUz gtYKES sc-hKgILt lhgqZA"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Retry
                </button>
              </styled.button>
            </BeagleButton>
          </ComponentWithTheme>
        </Styled(ComponentWithTheme)>
        <Styled(ComponentWithTheme)
          data-testid="show-more"
          onPress={[Function]}
          text="Show details ▼"
        >
          <ComponentWithTheme
            className="sc-eCssSg dkdlFW"
            data-testid="show-more"
            onPress={[Function]}
            text="Show details ▼"
          >
            <BeagleButton
              className="sc-eCssSg dkdlFW"
              data-testid="show-more"
              onPress={[Function]}
              text="Show details ▼"
            >
              <styled.button
                aria-disabled={false}
                className="sc-eCssSg dkdlFW"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                <button
                  aria-disabled={false}
                  className="sc-gsTCUz gtYKES sc-eCssSg dkdlFW"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Show details ▼
                </button>
              </styled.button>
            </BeagleButton>
          </ComponentWithTheme>
        </Styled(ComponentWithTheme)>
        <styled.section
          className=""
          data-testid="details"
        >
          <section
            className="sc-jSgupP cMxdgM"
            data-testid="details"
          >
            <p
              className="error-message"
              key="key_0"
            >
              <span
                className="error-indicator"
              >
                ►
              </span>
               ERROR:
               
              First Message
            </p>
            <p
              className="error-message"
              key="key_1"
            >
              <span
                className="error-indicator"
              >
                ►
              </span>
               ERROR:
               
              Second Message
            </p>
          </section>
        </styled.section>
      </section>
    </styled.section>
  </BeagleError>
</ComponentWithTheme>
`;

exports[`Beagle Error - render correctly without details Should match the snapshot 1`] = `
<ComponentWithTheme
  className="Test Class"
  errors={Array []}
  retry={[MockFunction]}
>
  <BeagleError
    className="Test Class"
    errors={Array []}
    retry={[MockFunction]}
  >
    <styled.section
      className="Test Class"
    >
      <section
        className="sc-dlfnbm DatSL Test Class"
      >
        <ComponentWithTheme
          text="Sorry!"
          textColor="red"
        >
          <BeagleText
            className=""
            text="Sorry!"
            textColor="red"
          >
            <styled.p
              className=""
              textColor="red"
            >
              <p
                className="sc-bdfBwQ eliSds"
              >
                Sorry!
              </p>
            </styled.p>
          </BeagleText>
        </ComponentWithTheme>
        <ComponentWithTheme
          text="An unexpected error happened while loading your page."
        >
          <BeagleText
            className=""
            text="An unexpected error happened while loading your page."
          >
            <styled.p
              className=""
            >
              <p
                className="sc-bdfBwQ fTPwPK"
              >
                An unexpected error happened while loading your page.
              </p>
            </styled.p>
          </BeagleText>
        </ComponentWithTheme>
        <Styled(ComponentWithTheme)
          data-testid="retry"
          onPress={[MockFunction]}
          text="Retry"
        >
          <ComponentWithTheme
            className="sc-hKgILt lhgqZA"
            data-testid="retry"
            onPress={[MockFunction]}
            text="Retry"
          >
            <BeagleButton
              className="sc-hKgILt lhgqZA"
              data-testid="retry"
              onPress={[MockFunction]}
              text="Retry"
            >
              <styled.button
                aria-disabled={false}
                className="sc-hKgILt lhgqZA"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                <button
                  aria-disabled={false}
                  className="sc-gsTCUz gtYKES sc-hKgILt lhgqZA"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Retry
                </button>
              </styled.button>
            </BeagleButton>
          </ComponentWithTheme>
        </Styled(ComponentWithTheme)>
      </section>
    </styled.section>
  </BeagleError>
</ComponentWithTheme>
`;
